rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,242 @@
+ * \param output_buffer is the data to write.
+ * \param uri is the relative file path to the output storage.
+ */
+void write_output_buffer(
+  const std::vector<uint8_t> & output_buffer,
+  const std::string & uri)
+{
+  const auto file_pointer = open_file(uri.c_str(), "wb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary writing!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  const auto nWrite = fwrite(
+    output_buffer.data(), sizeof(uint8_t),
+    output_buffer.size(), file_pointer);
+
+  if (nWrite != output_buffer.size()) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes written (" <<
+      nWrite << " != output_buffer_length (" << output_buffer.size() <<
+      ")!");
+    // An error indicator is set by fwrite, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to write decompressed data to file: \"" << uri << "\"!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  fclose(file_pointer);
+}
+
+/**
+ * Checks compression_result and throws a runtime_error if there was a ZSTD error.
+ */
+void throw_on_zstd_error(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD decompression error: " << ZSTD_getErrorName(compression_result);
+
+    throw std::runtime_error{error.str()};
+  }
+}
+
+/**
+ * Checks if the frame content is valid.
+ */
+void check_frame_content(const size_t frame_content)
+{
+  if (frame_content == ZSTD_CONTENTSIZE_ERROR) {
+    throw std::runtime_error{"Unable to determine file size due to error."};
+  } else if (frame_content == ZSTD_CONTENTSIZE_UNKNOWN) {
+    throw std::runtime_error{"Unable to determine file size."};
+  }
+}
+
+/**
+ * Prints decompression statistics to the debug log stream.
+ * The log statement is formatted as JSON.
+ *
+ * Example:
+ *  "Decompression statistics" : {"Time" : 1.2, "Compression Ratio" : 0.5}
+ *
+ * \param start is the time_point when compression started.
+ * \param end is the time_point when compression ended.
+ * \param decompressed_size is the file size after decompression
+ * \param compressed_size is the compressed file size
+ */
+void print_decompression_statistics(
+  const std::chrono::high_resolution_clock::time_point start,
+  const std::chrono::high_resolution_clock::time_point end,
+  const size_t decompressed_size,
+  const size_t compressed_size)
+{
+  const auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
+  const auto decompression_ratio =
+    static_cast<double>(decompressed_size) / static_cast<double>(compressed_size);
+
+  ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+    "\"Decompression statistics\" : {" <<
+      "\"Time\" : " << (duration.count() / 1000.0) <<
+      ", \"Compression Ratio\" : " << decompression_ratio <<
```

Comment:
`Decompression Ratio` and not Compression? Does it make sense to keep this though?